scheduling/taskmanager.go

```diff
@@ -0,0 +1,415 @@
+	"strings"
+	"time"
+)
+
+type Key interface {
+	String() string
+}
+
+type TaskMetadataKey struct {
+	taskID TaskID
+}
+
+func (tmk TaskMetadataKey) String() string {
+	if tmk.taskID == 0 {
+		return "/tasks/metadata/task_id="
+	}
+	return fmt.Sprintf("/tasks/metadata/task_id=%d", tmk.taskID)
+}
+
+type TaskRunKey struct {
+	taskID TaskID
+}
+
+func (trk TaskRunKey) String() string {
+	if trk.taskID == 0 {
+		return "/tasks/runs/task_id="
+	}
+	return fmt.Sprintf("/tasks/runs/task_id=%d", trk.taskID)
+}
+
+type TaskRunMetadataKey struct {
+	taskID TaskID
+	runID  TaskRunID
+	date   time.Time
+}
+
+func (trmk TaskRunMetadataKey) String() string {
+	key := "/tasks/runs/metadata"
+
+	// adds the date to the key if it's not the default value
+	if trmk.date != (time.Time{}) {
+		key += fmt.Sprintf("/%s", trmk.date.Format("2006/01/02"))
+	}
+
+	// adds the task_id and run_id to the key if they're not the default value
+	if trmk.taskID != 0 && trmk.runID != 0 {
+		key += fmt.Sprintf("/task_id=%d/run_id=%d", trmk.taskID, trmk.runID)
+	}
+	return key
+}
+
+func NewTaskManager(storage StorageProvider) TaskManager {
+	return TaskManager{storage: storage}
+}
+
+type TaskManager struct {
+	storage StorageProvider
+}
+
+type TaskMetadataList []TaskMetadata
+
+func (tml *TaskMetadataList) ToJSON() string {
+	return ""
+}
+
+// Task Methods
+func (tm *TaskManager) CreateTask(name string, tType TaskType, target TaskTarget) (TaskMetadata, error) {
+	keys, err := tm.storage.ListKeys(TaskMetadataKey{}.String())
+	if err != nil {
+		return TaskMetadata{}, fmt.Errorf("failed to fetch keys: %v", err)
+	}
+
+	// This logic could probably be somewhere else
+	var latestID int
+	if len(keys) == 0 {
+		latestID = 0
+	} else {
+		latestID, err = getLatestID(keys)
+		if err != nil {
+			return TaskMetadata{}, err
+		}
+	}
+
+	metadata := TaskMetadata{
+		ID:          TaskID(latestID + 1),
+		Name:        name,
+		TaskType:    tType,
+		Target:      target,
+		TargetType:  target.Type(),
+		DateCreated: time.Now().UTC(),
+	}
+
+	// I do this serialize and deserialize a lot in this file. Would be nice to have set and get helpers that deal with
+	// all the converting instead
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		return TaskMetadata{}, fmt.Errorf("failed to marshal metadata: %v", err)
+	}
+	err = tm.storage.Set(TaskMetadataKey{taskID: metadata.ID}.String(), string(serializedMetadata))
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	runs := TaskRuns{
+		TaskID: metadata.ID,
+		Runs:   []TaskRunSimple{},
+	}
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	taskRunKey := TaskRunKey{taskID: metadata.ID}
+	err = tm.storage.Set(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+// Finds the highest increment in a list of strings formatted like "/tasks/metadata/task_id=0"
+func getLatestID(taskPaths []string) (int, error) {
```

Comment:
will accidental dupes be an issue? wondering if concurrent calls might result in the same IDs getting returned.